File: lab2/gr9sub1/ЛУЦКИЙ_ВАДИМ/src/supervisor.py
#!/usr/bin/env python3
import os
import sys
import time
import signal
import json
from datetime import datetime, timedelta

class Supervisor:
    def __init__(self, config_file):
        self.config_file = config_file
        self.workers = {}  # PID -> worker_info
        self.running = True
        self.restart_history = []  # История рестартов для ограничения частоты
        
        # Загрузка конфигурации
        self.load_config()
        
        # Установка обработчиков сигналов
        signal.signal(signal.SIGTERM, self.signal_handler)
        signal.signal(signal.SIGINT, self.signal_handler)
        signal.signal(signal.SIGHUP, self.signal_handler)
        signal.signal(signal.SIGUSR1, self.signal_handler)
        signal.signal(signal.SIGUSR2, self.signal_handler)
        signal.signal(signal.SIGCHLD, self.signal_handler)
    
    def load_config(self):
        """Загрузка конфигурации"""
        try:
            with open(self.config_file, 'r') as f:
                self.config = json.load(f)
            print("Configuration loaded successfully")
        except Exception as e:
            print(f"Error loading config: {e}")
            sys.exit(1)
    
    def signal_handler(self, signum, frame):
        """Обработчик сигналов"""
        if signum in (signal.SIGTERM, signal.SIGINT):
            print(f"Supervisor (PID {os.getpid()}): Received shutdown signal")
            self.graceful_shutdown()
        elif signum == signal.SIGHUP:
            print(f"Supervisor (PID {os.getpid()}): Received reload signal")
            self.graceful_reload()
        elif signum == signal.SIGUSR1:
            print(f"Supervisor (PID {os.getpid()}): Received SIGUSR1, switching workers to light mode")
            self.broadcast_signal(signal.SIGUSR1)
        elif signum == signal.SIGUSR2:
            print(f"Supervisor (PID {os.getpid()}): Received SIGUSR2, switching workers to heavy mode")
            self.broadcast_signal(signal.SIGUSR2)
        elif signum == signal.SIGCHLD:
            # Обработка завершения дочерних процессов
            self.handle_child_exit()
    
    def can_restart(self):
        """Проверка возможности рестарта (не более 5 за 30 секунд)"""
        now = datetime.now()
        # Удаляем старые записи (старше 30 секунд)
        self.restart_history = [t for t in self.restart_history if now - t < timedelta(seconds=30)]
        
        # Проверяем, что не превышен лимит
        return len(self.restart_history) < 5
    
    def handle_child_exit(self):
        """Обработка завершения дочерних процессов"""
        try:
            while True:
                # Неблокирующий waitpid для сбора зомби
                pid, status = os.waitpid(-1, os.WNOHANG)
                if pid == 0:  # Нет завершённых процессов
                    break
                
                if pid in self.workers:
                    worker_info = self.workers[pid]
                    print(f"Worker {worker_info['id']} (PID {pid}) exited with status {status}")
                    
                    # Удаляем из активных процессов
                    del self.workers[pid]
                    
                    # Перезапуск, если супервизор ещё работает
                    if self.running and self.can_restart():
                        print(f"Restarting worker {worker_info['id']}")
                        self.restart_history.append(datetime.now())
                        self.start_worker(worker_info['id'])
                    else:
                        print(f"Restart limit exceeded for worker {worker_info['id']}")
                
        except ChildProcessError:
            pass  # Нет дочерних процессов
    
    def start_worker(self, worker_id):
        """Запуск воркера"""
        try:
            # Запускаем воркер как дочерний процесс
            pid = os.fork()
            
            if pid == 0:  # Дочерний процесс
                # Запускаем воркер
                os.execlp(sys.executable, sys.executable, 'worker.py', 
                         str(worker_id), self.config_file)
            else:  # Родительский процесс
                self.workers[pid] = {'id': worker_id, 'start_time': time.time()}
                print(f"Started worker {worker_id} with PID {pid}")
                
        except Exception as e:
            print(f"Error starting worker {worker_id}: {e}")
    
    def broadcast_signal(self, sig):
        """Отправка сигнала всем воркерам"""
        for pid in list(self.workers.keys()):
            try:
                os.kill(pid, sig)
                print(f"Sent signal {sig} to worker PID {pid}")
            except ProcessLookupError:
                # Процесс уже завершился
                if pid in self.workers:
                    del self.workers[pid]
    
    def graceful_shutdown(self):
        """Корректное завершение работы"""
        print("Initiating graceful shutdown...")
        self.running = False
        
        # Отправляем SIGTERM всем воркерам
        self.broadcast_signal(signal.SIGTERM)
        
        # Ждём завершения воркеров (до 5 секунд)
        start_time = time.time()
        while self.workers and (time.time() - start_time < 5):
            time.sleep(0.1)
            # Проверяем завершившиеся процессы
            self.handle_child_exit()
        
        # Если остались работающие воркеры - отправляем SIGKILL
        if self.workers:
            print("Forcing termination of remaining workers...")
            for pid in list(self.workers.keys()):
                try:
                    os.kill(pid, signal.SIGKILL)
                except ProcessLookupError:
                    pass
        
        print("Supervisor shutdown complete")
        sys.exit(0)
    
    def graceful_reload(self):
        """Graceful reload конфигурации"""
        print("Initiating graceful reload...")
        
        # Перезагружаем конфигурацию
        old_config = self.config.copy()
        self.load_config()
        
        # Отправляем SIGTERM старым воркерам
        self.broadcast_signal(signal.SIGTERM)
        
        # Ждём их завершения
        start_time = time.time()
        while self.workers and (time.time() - start_time < 5):
            time.sleep(0.1)
            self.handle_child_exit()
        
        # Запускаем новых воркеров с новой конфигурацией
        for i in range(self.config['workers']):
            self.start_worker(i + 1)
        
        print("Graceful reload complete")
    
    def run(self):
        """Основной цикл супервизора"""
        print(f"Supervisor (PID {os.getpid()}) starting {self.config['workers']} workers...")
        
        # Запуск воркеров
        for i in range(self.config['workers']):
            self.start_worker(i + 1)
        
        print("Supervisor running. Send signals to PID", os.getpid())
        
        # Основной цикл
        while self.running:
            time.sleep(1)
            
            # Периодическая проверка состояния
            if len(self.workers) < self.config['workers'] and self.running:
                # Запускаем недостающих воркеров
                active_ids = {info['id'] for info in self.workers.values()}
                for i in range(1, self.config['workers'] + 1):
                    if i not in active_ids and self.can_restart():
                        print(f"Starting missing worker {i}")
                        self.restart_history.append(datetime.now())
                        self.start_worker(i)

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: supervisor.py <config_file>")
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    supervisor = Supervisor(config_file)
    supervisor.run()

File: lab2/gr9sub1/TITKOVA_ANNA/src/supervisor/supervisor.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"sync"
	"syscall"
	"time"
)

type Config struct {
	Workers     int               `json:"workers"`
	ModeDefault string            `json:"mode_default"`
	ModeHeavy   ModeConfig        `json:"mode_heavy"`
	ModeLight   ModeConfig        `json:"mode_light"`
	WorkerAttrs []WorkerAttr      `json:"worker_attrs"`
	NiceMap     map[string]int    `json:"nice_map"`
	AffinityMap map[string]string `json:"affinity_map"`
}

type ModeConfig struct {
	WorkUS  int `json:"work_us"`
	SleepUS int `json:"sleep_us"`
}

type WorkerAttr struct {
	Nice int    `json:"nice"`
	CPUs string `json:"cpus"`
}

type Supervisor struct {
	config       *Config
	workers      map[int]*exec.Cmd
	restartTimes map[int][]time.Time
	mutex        sync.Mutex
	terminate    bool
	reload       bool
	switchMode   string
}

func NewSupervisor(configPath string, workers int) (*Supervisor, error) {
	config, err := loadConfig(configPath)
	if err != nil {
		return nil, err
	}

	if workers > 0 {
		config.Workers = workers
	}

	return &Supervisor{
		config:       config,
		workers:      make(map[int]*exec.Cmd),
		restartTimes: make(map[int][]time.Time),
	}, nil
}

func loadConfig(path string) (*Config, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read config file: %w", err)
	}

	var config Config
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, fmt.Errorf("failed to parse config JSON: %w", err)
	}

	// Валидация конфига
	if config.Workers < 2 {
		config.Workers = 2
		log.Printf("Warning: workers count increased to minimum: %d", config.Workers)
	}

	return &config, nil
}

func (s *Supervisor) Start() error {
	// Настройка обработчиков сигналов
	s.setupSignals()

	log.Printf("Starting %d workers", s.config.Workers)

	// Запуск начальных воркеров
	for i := 0; i < s.config.Workers; i++ {
		s.startWorker(i)
	}

	return s.mainLoop()
}

func (s *Supervisor) setupSignals() {
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh,
		syscall.SIGTERM, // graceful shutdown
		syscall.SIGINT,  // Ctrl+C
		syscall.SIGHUP,  // reload config
		syscall.SIGUSR1, // switch to light mode
		syscall.SIGUSR2, // switch to heavy mode
		syscall.SIGCHLD, // child process status change
	)

	go func() {
		for sig := range sigCh {
			switch sig {
			case syscall.SIGTERM, syscall.SIGINT:
				log.Printf("Received %v -> initiating graceful shutdown", sig)
				s.terminate = true
			case syscall.SIGHUP:
				log.Println("Received SIGHUP -> reloading configuration")
				s.reload = true
			case syscall.SIGUSR1:
				log.Println("Received SIGUSR1 -> switching all workers to LIGHT mode")
				s.switchMode = "light"
			case syscall.SIGUSR2:
				log.Println("Received SIGUSR2 -> switching all workers to HEAVY mode")
				s.switchMode = "heavy"
			case syscall.SIGCHLD:
				s.reapWorkers()
			}
		}
	}()
}

func (s *Supervisor) startWorker(id int) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	// Проверка ограничения частоты рестартов
	now := time.Now()
	times := s.restartTimes[id]

	// Удаляем старые записи (старше 30 секунд)
	var recent []time.Time
	for _, t := range times {
		if now.Sub(t) <= 30*time.Second {
			recent = append(recent, t)
		}
	}

	// Проверяем лимит (не более 5 рестартов за 30 секунд)
	if len(recent) >= 5 {
		log.Printf("Worker %d restart rate limit exceeded (%d restarts in 30s) - skipping restart", id, len(recent))
		return
	}

	// Построение команды для воркера
	workerPath := "./worker"
	if _, err := os.Stat(workerPath); os.IsNotExist(err) {
		// Если worker нет в текущей директории, ищем рядом с супервизором
		workerPath = filepath.Join(filepath.Dir(os.Args[0]), "worker")
	}

	cmd := exec.Command(workerPath,
		"--id", fmt.Sprintf("%d", id),
		"--config", "config/config.json")

	// Применяем атрибуты воркера
	attrs := s.getWorkerAttrs(id)
	if attrs.Nice != 0 {
		cmd.Args = append(cmd.Args, "--nice", fmt.Sprintf("%d", attrs.Nice))
	}
	if attrs.CPUs != "" {
		cmd.Args = append(cmd.Args, "--cpus", attrs.CPUs)
	}

	// Настраиваем стандартные потоки
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		log.Printf("Failed to start worker %d: %v", id, err)
		return
	}

	s.workers[id] = cmd
	s.restartTimes[id] = append(recent, now)

	log.Printf("Started worker %d with PID %d (nice=%d, cpus=%s)",
		id, cmd.Process.Pid, attrs.Nice, attrs.CPUs)
}

func (s *Supervisor) getWorkerAttrs(id int) WorkerAttr {
	// Сначала проверяем worker_attrs массив
	if id < len(s.config.WorkerAttrs) {
		return s.config.WorkerAttrs[id]
	}

	// Fallback к мапам
	attr := WorkerAttr{}
	idStr := fmt.Sprintf("%d", id)
	if nice, ok := s.config.NiceMap[idStr]; ok {
		attr.Nice = nice
	}
	if cpus, ok := s.config.AffinityMap[idStr]; ok {
		attr.CPUs = cpus
	}

	return attr
}

func (s *Supervisor) reapWorkers() {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	// Используем неблокирующий wait для сбора всех завершенных детей
	for {
		var status syscall.WaitStatus
		pid, err := syscall.Wait4(-1, &status, syscall.WNOHANG, nil)
		if err != nil || pid <= 0 {
			break // Нет больше завершенных процессов
		}

		// Находим какой воркер завершился
		for id, cmd := range s.workers {
			if cmd.Process != nil && cmd.Process.Pid == pid {
				log.Printf("Worker %d (PID %d) exited with status %d",
					id, pid, status.ExitStatus())

				delete(s.workers, id)
				if !s.terminate {
					log.Printf("Restarting worker %d", id)
					go s.startWorker(id)
				}
				break
			}
		}
	}
}

func (s *Supervisor) broadcastSignal(sig syscall.Signal) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	log.Printf("Broadcasting signal %v to all workers", sig)
	count := 0
	for id, cmd := range s.workers {
		if cmd.Process != nil {
			log.Printf("Sending signal %v to worker %d (PID %d)", sig, id, cmd.Process.Pid)
			if err := cmd.Process.Signal(sig); err != nil {
				log.Printf("Failed to send signal %v to worker %d: %v", sig, id, err)
			} else {
				count++
			}
		}
	}
	log.Printf("Sent signal %v to %d workers", sig, count)
}

func (s *Supervisor) gracefulShutdown() {
	log.Println("Initiating graceful shutdown...")

	// Отправляем SIGTERM всем воркерам
	s.broadcastSignal(syscall.SIGTERM)

	// Ждем завершения процессов
	timeout := time.After(5 * time.Second)
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	log.Println("Waiting for workers to exit gracefully...")

	for range ticker.C {
		allStopped := true
		s.mutex.Lock()
		for id, cmd := range s.workers {
			if cmd.Process != nil {
				if err := cmd.Process.Signal(syscall.Signal(0)); err == nil {
					allStopped = false
					log.Printf("Worker %d (PID %d) still running...", id, cmd.Process.Pid)
				} else {
					// Процесс завершился
					delete(s.workers, id)
				}
			}
		}
		s.mutex.Unlock()

		if allStopped {
			log.Println("All workers exited gracefully")
			break
		}

		select {
		case <-timeout:
			log.Println("Timeout reached - forcing shutdown of remaining workers")
			s.broadcastSignal(syscall.SIGKILL)
			return
		default:
		}
	}

	log.Println("Graceful shutdown completed")
}

func (s *Supervisor) reloadConfig() {
    log.Println("=== RELOAD CONFIGURATION ===")
    
    // ОТЛАДКА: покажем текущие PID до reload
    s.mutex.Lock()
    currentPIDs := make(map[int]int)
    for i, cmd := range s.workers {
        if cmd.Process != nil {
            currentPIDs[i] = cmd.Process.Pid
        }
    }
    s.mutex.Unlock()
    log.Printf("BEFORE RELOAD - Current worker PIDs: %v", currentPIDs)

    // 1. Загрузить новую конфигурацию (ИСПРАВЛЕННЫЙ ПУТЬ)
    configPath := "config/config.json" // Правильный путь к конфигу
    log.Printf("Loading config from: %s", configPath)
    
    config, err := loadConfig(configPath)
    if err != nil {
        log.Printf("Failed to reload config from %s: %v", configPath, err)
        return
    }

    log.Printf("New configuration: workers=%d (old=%d), mode_default=%s",
        config.Workers, len(currentPIDs), config.ModeDefault)

    // 2. АГРЕССИВНО убиваем всех текущих воркеров
    s.mutex.Lock()
    log.Printf("Killing all %d workers", len(s.workers))
    
    for i, cmd := range s.workers {
        if cmd.Process != nil {
            oldPID := cmd.Process.Pid
            log.Printf("Sending SIGKILL to worker %d (PID %d)", i, oldPID)
            // Немедленное убийство, не ждем graceful shutdown
            cmd.Process.Signal(syscall.SIGKILL)
        }
    }
    
    // ПОЛНОСТЬЮ очищаем мапы
    s.workers = make(map[int]*exec.Cmd)
    s.restartTimes = make(map[int][]time.Time)
    s.mutex.Unlock()

    // 3. Ждем чтобы процессы точно умерли
    log.Println("Waiting for processes to die...")
    time.Sleep(1 * time.Second)

    // 4. Запускаем НОВЫХ воркеров
    log.Printf("Starting %d NEW workers", config.Workers)
    newPIDs := make(map[int]int)
    
    for i := 0; i < config.Workers; i++ {
        s.startWorker(i)
        
        // Собираем новые PID
        s.mutex.Lock()
        if cmd, exists := s.workers[i]; exists && cmd.Process != nil {
            newPIDs[i] = cmd.Process.Pid
            log.Printf("Started worker %d with NEW PID: %d", i, cmd.Process.Pid)
        }
        s.mutex.Unlock()
    }

    // 5. Детальная проверка
    log.Printf("=== RELOAD COMPLETED ===")
    log.Printf("Old PIDs: %v", currentPIDs)
    log.Printf("New PIDs: %v", newPIDs)
    
    // Проверяем изменения
    changed := false
    for i := range newPIDs {
        if oldPID, wasRunning := currentPIDs[i]; wasRunning {
            if newPIDs[i] != oldPID {
                changed = true
                log.Printf("PID CHANGED: worker %d: %d -> %d", i, oldPID, newPIDs[i])
            } else {
                log.Printf("PID UNCHANGED: worker %d: %d", i, oldPID)
            }
        }
    }
    
    if changed {
        log.Println("SUCCESS: Worker PIDs changed after reload!")
    } else {
        log.Println("PROBLEM: Worker PIDs did not change - reload may not be working")
    }
}

func (s *Supervisor) mainLoop() error {
	ticker := time.NewTicker(200 * time.Millisecond)
	defer ticker.Stop()

	log.Println("Supervisor main loop started")

	for {
		select {
		case <-ticker.C:
			// Проверка умерших воркеров (fallback)
			s.reapWorkers()

			// Обработка переключения режимов
			if s.switchMode != "" {
				var sig syscall.Signal
				if s.switchMode == "light" {
					sig = syscall.SIGUSR1
				} else {
					sig = syscall.SIGUSR2
				}
				log.Printf("Broadcasting %s mode to all workers", s.switchMode)
				s.broadcastSignal(sig)
				s.switchMode = ""
			}

			// Обработка перезагрузки конфигурации
			if s.reload {
				s.reloadConfig()
				s.reload = false
			}

			// Обработка завершения работы
			if s.terminate {
				s.gracefulShutdown()
				return nil
			}
		}
	}
}
